File: flutter_app/lib/domain/rounds.dart
import 'package:uuid/uuid.dart';

import 'constants.dart';
import 'models.dart';

const _uuid = Uuid();

class CreateRoundInput {
  const CreateRoundInput({
    required this.playedAt,
    required this.courseId,
    required this.courseName,
    required this.roundType,
    required this.configSnapshot,
    this.holePars,
  });

  final String playedAt;
  final String? courseId;
  final String courseName;
  final RoundType roundType;
  final RoundConfig configSnapshot;
  final List<int>? holePars;
}

RoundRecord createRoundRecord(CreateRoundInput input) {
  final now = DateTime.now().toIso8601String();
  final id = _uuid.v4();
  final normalizedName =
      input.courseName.trim().isEmpty ? '未設定コース' : input.courseName.trim();
  final holePars = input.holePars ?? parTemplates[input.roundType]!;

  final round = Round(
    id: id,
    playedAt: input.playedAt,
    startedAt: now,
    finishedAt: null,
    status: RoundStatus.inProgress,
    courseId: input.courseId,
    courseName: normalizedName,
    roundType: input.roundType,
    holesCount: holePars.length,
    holePars: holePars,
    currentHoleNo: 1,
    configSnapshot: input.configSnapshot,
    totalScore: null,
    lastInputAt: now,
  );

  return RoundRecord(
    round: round,
    holes: [
      for (var i = 0; i < holePars.length; i++)
        createEmptyHole(round.id, i + 1, holePars[i]),
    ],
  );
}

HoleScore createEmptyHole(String roundId, int holeNo, int par) {
  return HoleScore(
    roundId: roundId,
    holeNo: holeNo,
    par: par,
    strokes: null,
    putts: null,
    teeShotDirection: null,
    puttsUnknown: false,
    penaltyTotal: 0,
    obCount: 0,
    otherPenaltyCount: 0,
    fwKeep: null,
    gir: null,
    bunkerIn: null,
    bunkerCount: 0,
    updatedAt: DateTime.now().toIso8601String(),
  );
}

RoundConfig mergeConfig([RoundConfig? config]) {
  if (config == null) return defaultRoundConfig;
  return defaultRoundConfig.copyWith(
    fw: config.fw,
    gir: config.gir,
    ob: config.ob,
    bunker: config.bunker,
    puttUnknown: config.puttUnknown,
  );
}

File: lib/domain/rounds.ts
import { z } from "zod";
import { DEFAULT_ROUND_CONFIG, PAR_TEMPLATES } from "@/lib/domain/constants";
import type { HoleScore, Round, RoundConfig, RoundRecord, RoundType } from "@/lib/domain/types";

const roundTypeSchema = z.enum(["FULL_18", "HALF_9", "SHORT"]);

export const createRoundInputSchema = z.object({
  playedAt: z.string().min(1),
  courseId: z.string().nullable().optional(),
  courseName: z.string().max(80),
  roundType: roundTypeSchema,
  configSnapshot: z.object({
    fw: z.boolean(),
    gir: z.boolean(),
    ob: z.boolean(),
    bunker: z.boolean(),
    puttUnknown: z.boolean(),
  }),
  holePars: z.array(z.number().int().min(3).max(6)).optional(),
});

export type CreateRoundInput = z.infer<typeof createRoundInputSchema>;

export function createRoundRecord(input: CreateRoundInput): RoundRecord {
  const now = new Date().toISOString();
  const id = crypto.randomUUID();
  const parsed = createRoundInputSchema.parse(input);
  const holePars = parsed.holePars ?? PAR_TEMPLATES[parsed.roundType];

  const round: Round = {
    id,
    playedAt: parsed.playedAt,
    startedAt: now,
    finishedAt: null,
    status: "IN_PROGRESS",
    courseId: parsed.courseId ?? null,
    courseName: parsed.courseName.trim(),
    roundType: parsed.roundType,
    holesCount: holePars.length,
    holePars,
    currentHoleNo: 1,
    configSnapshot: parsed.configSnapshot,
    totalScore: null,
    lastInputAt: now,
  };

  return {
    round,
    holes: holePars.map((par, index) => createEmptyHole(round.id, index + 1, par)),
  };
}

export function createEmptyHole(roundId: string, holeNo: number, par: number): HoleScore {
  return {
    roundId,
    holeNo,
    par,
    strokes: null,
    putts: null,
    teeShotDirection: null,
    puttsUnknown: false,
    penaltyTotal: 0,
    obCount: 0,
    otherPenaltyCount: 0,
    fwKeep: null,
    gir: null,
    bunkerIn: null,
    updatedAt: new Date().toISOString(),
  };
}

export function mergeConfig(config?: Partial<RoundConfig>): RoundConfig {
  return {
    ...DEFAULT_ROUND_CONFIG,
    ...config,
  };
}

export function roundTypeFromLabel(value: string): RoundType {
  if (value === "HALF_9" || value === "SHORT") return value;
  return "FULL_18";
}
